python/cudf/cudf/core/indexed_frame.py

```diff
@@ -2898,22 +2977,32 @@ def _drop_na_rows(self, how="any", subset=None, thresh=None):
             self._index.names,
         )
 
-    def _apply_boolean_mask(self, boolean_mask):
+    def _apply_boolean_mask(
+        self, boolean_mask, keep_index=True, normalize_and_check=True
+    ):
         """Apply boolean mask to each row of `self`.
 
         Rows corresponding to `False` is dropped.
+
+        If keep_index is False, the index is not preserved.
         """
-        boolean_mask = cudf.core.column.as_column(boolean_mask)
-        if not is_bool_dtype(boolean_mask.dtype):
-            raise ValueError("boolean_mask is not boolean type.")
-        if (bn := len(boolean_mask)) != (n := len(self)):
-            raise IndexError(f"Boolean mask has wrong length: {bn} not {n}")
+        if normalize_and_check:
```

Comment:
I think this would be helpful. I especially think the term `normalized` is being overloaded quite a bit and we need to either hash out terminology or use more typing to be able to distinguish between:

* the kind of normalization being done by `unpack_iloc_key`
* the kind of normalization being done by `normalize_row_iloc_indexer`
* other kind of normalization?

Reply:
Let me introduce types for gather maps and boolean masks separately (then the `check_bounds` arguments can disappear here entirely).